### ISSUE TITLE
v5 wish list

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#2403]
Do you miss a feature in WebdriverIO? Do you have any feedback how we can make it better? Let us know and help use prioritize features for a new major release!

**Comment by KenzoM** [comment]
@christian-bromann What are some of the proposed features so far?

**Comment by christian-bromann** [comment]
@KenzoM I haven't been through all tickets with my triage but I made a [`v5`](https://github.com/webdriverio/webdriverio/issues?q=is%3Aopen+is%3Aissue+milestone%3Av5) milestone

**Comment by TuHuynhVan** [comment]
@christian-bromann 
I would like to see this request comes true:
https://github.com/webdriverio/webdriverio/issues/1695

**Comment by ORESoftware** [comment]
+1

**Comment by mkj28** [comment]
I would _love_ to see test retry for Jasmine....

**Comment by christian-bromann** [comment]
@mkj28 have you seen http://webdriver.io/guide/testrunner/retry.html#Rerun-single-tests-in-Jasmine-or-Mocha

**Comment by Efefefef** [comment]
I would love to have
https://github.com/webdriverio/webdriverio/issues/2185
https://github.com/webdriverio/webdriverio/issues/1890

**Comment by ORESoftware** [comment]
@mkj28 what do you mean by retry exactly? what API would you like to see? Or link to an existing issue

**Comment by dguzmanSM** [comment]
Hey Cris, I try this awhile ago, and was still an issue for me: https://github.com/webdriverio/wdio-cucumber-framework/issues/36

**Comment by KenzoM** [comment]
Can we also integrate axe-core for Web Accessibility?
@klamping was awesome to show us the basic setup. It would be neat if V5 comes with it.

https://www.youtube.com/watch?v=6gwtqYCxNwQ&t=26s

**Comment by christian-bromann** [comment]
@KenzoM well axe-core already works with WebdriverIO it is just not simplified to one command. After all this would be a plugin and nothing that would be integrated into WebdriverIO core.

**Comment by klamping** [comment]
@KenzoM fyi, someone has taken a stab at putting together an Axe service here: https://github.com/cerner/terra-toolkit/blob/master/src/wdio/services/AxeService.js

**Comment by klamping** [comment]
One idea I have is to prompt for Sauce/BS/TB credentials if they aren't found (but asked for in the config). Right now, when you first initialize the config, it prompts you for your credentials. But once that's set it, it assumes it's already there.

The problem is, if someone new joins the project, they have to manually set up those Env variables. It's not a big task, but can be confusing to newcomers and those less familiar with env variables. 

Instead, have WDIO do a detection if SL/BS/TB is being used (maybe if the service is set?), and prompt for (and set) those env variables for the user.

**Comment by sherlock1982** [comment]
There are lots of commands which have variants. You can find them with pattern /did not match a known command/. I don't know why but for some reason priority is given to commands which are not W3C compatible. 

For example `browser.execute` and `browser.windowHandlePosition` always takes two requests on standalone geckodriver. 

Maybe it makes sense to try W3C commands first?

**Comment by christian-bromann** [comment]
> Maybe it makes sense to try W3C commands first?

When we release v5 most of the driver probably already speak the latest W3C spec and we can switch the order. Good point! Right now I think it only affects FF tests.

**Comment by KenzoM** [comment]
What about having Jest Support?
https://github.com/webdriverio/webdriverio/issues/2052

**Comment by erwinheitzman** [comment]
I would like to see even better support for page objects.
Currently some methods do not work as expected when coupled with page objects such as the scroll command which is made clear in the following testcases that I created:

```
    it('should scroll to specific element while chained', async function () {
        await this.client.windowHandleSize({width: 100, height: 100})
        await this.client.element('.box').scroll()

        const scrollPos = await this.client.execute(() => {
            return {
                x: $(document).scrollLeft(),
                y: $(document).scrollTop()
            }
        })

        scrollPos.value.x.should.be.approximately(15, 15)
        scrollPos.value.y.should.be.approximately(262, 40)
    })
```

```
    it('should scroll to specific element with offset while chained', async function () {
        await this.client.windowHandleSize({width: 100, height: 100})
        await this.client.element('.box').scroll(-10, -22)

        const scrollPos = await this.client.execute(() => {
            return {
                x: $(document).scrollLeft(),
                y: $(document).scrollTop()
            }
        })

        scrollPos.value.x.should.be.approximately(5, 15)
        scrollPos.value.y.should.be.approximately(240, 40)
    })
```

**Comment by christian-bromann** [comment]
@erwinheitzman not sure why scrolling is important in your test, it usually isn't. Can you explain in more detail what the difference between:
```js
await this.client.element('.box').scroll()
```
and
```
await this.client.element('.box').scroll(-10, -22)
```
is and what lack of support you see here?

**Comment by erwinheitzman** [comment]
@christian-bromann I agree that it usually isn't needed but when I was having trouble clicking a specific element a few days ago I tried scrolling before clicking.
When I tried PageObject.myGetter.scroll() I was suprised it didn't work the same as browser.scroll('my-selector').

The difference between the given examples is that one does not use any offset and the other does, I was trying to see if I could fix the behaviour for the scroll method so I added those tests locally (never got it working though as one test kept failing).

The support I was missing is to be able to use MyPageObject.myGetter.scroll() with and without offsets and with and without a selector.

It's not a absolute must have but I think it would be a nice thing if we could support it. Perhaps it's a nice addition for a version further down the line.

**Comment by christian-bromann** [comment]
> Perhaps it's a nice addition for a version further down the line.

Probably. I think it is important to provide consistent behavior. The issue is known and similar to #1879. I guess this will get definitely fixed within the next iterations.

However issues with clicks are common. The reason is usually because there is another element (e.g. with `position:fixed` located on top of the target element. I would rather remove that element instead of using scoll. Ultimately the scrolling command will be removed.

**Comment by erwinheitzman** [comment]
I'd love to see [#2357](https://github.com/webdriverio/webdriverio/issues/2357) get implemented :)

**Comment by mkj28** [comment]
> what do you mean by retry exactly

@ORESoftware currently you can only rerun test block or hook, not entire test or test suite. So what I mean is having a global (say in `wdio.conf.js`) number defined of retries for entire test. Lower priority (but still better than nothing) would be rerunning entire suite.

Currently the only way to achieve this with Jasmine I can see is writing my own reporter or parsing test results myself and doing another webdriver.io pass for the specific test suites.

**Comment by kevinmcdonnell** [comment]
I think some performance profiling around running tests in parallel would be interesting. And if there could be any gains there

https://github.com/webdriverio/webdriverio/issues/1643

**Comment by kevinmcdonnell** [comment]
Also if I'm trying to fix a flaky test I like to rerun it x number of times to see if it passes consistently. Currently I just wrap the test in a for loop, not sure what a nice syntax would look like though, maybe similar to wdio implementation of retry flaky tests

**Comment by kevinmcdonnell** [comment]
@christian-bromann have you thought about how wdio could use lambda functions to execute tests?

**Comment by christian-bromann** [comment]
> lambda functions to execute tests?

@kevinmcdonnell not sure what you mean by lambda functions. Are you talking about AWS lambda? If so I guess it is an interesting approach to run even more tests in parallel.

**Comment by kevinmcdonnell** [comment]
Yeah there's been a load of projects popping up around running chrome (https://www.npmjs.com/package/@serverless-chrome/lambda) and tests in lambda.

This looks like a very interesting project for example https://github.com/graphcool/chromeless#api-documentation

Clay Smith here has gotten node tests with webdriver working in a lambda https://medium.com/clog/running-selenium-and-headless-chrome-on-aws-lambda-fb350458e4df

running tests on lambda is a game changer for running tests quickly. A 20 min suite could simply be as slow as the slowest test

**Comment by dguzmanSM** [comment]
Hey guys, what is an ETA for v5 release?

**Comment by christian-bromann** [comment]
@dguzmanSM I work on this on my freetime, so whenever I get the time for it, probably not this year, please follow us on Twitter and don't ask for ETAs

**Comment by dguzmanSM** [comment]
@christian-bromann sorry man, didn't want to sound pushy, I know its a open source project, and it is AWESOME!, just talked about WebdriverIO in first TestingCR group MeetUp, here in Costa Rica. People love it!

**Comment by kevinmcdonnell** [comment]
just an update on my lambda idea. It looks like this will be more easily achieved using AWS Fargate (https://aws.amazon.com/fargate/ is a new product still in beta). 

The idea is our jenkins workflow will create a container which has all the test dependencies and a checkout of the code. If our tests suite has 100 spec files we use fargate to run 100 containers which will each run 1 spec file. We would need some way to orchestrate communication with fargate and the division of spec files, I guess this could be achieved by a wdio plugin (need to look up how to create one)

**Comment by christian-bromann** [comment]
@kevinmcdonnell this is a great idea and will probably get shipped within the new test runner of v5. I guess the simplest integration is to check for an AWS or GCE key and run them automatically as lambda function if they exist in the config.

**Comment by kevinmcdonnell** [comment]
Note this is a new product AWS fargate not AWS lambda. Lambda has limitation on the size of what you can pass to it (which could possibly be made to work but may be harder and less versatile) whereas fargate just executes a container that you've built so it can easily have dependencies like chrome/firefox, choice of OS, a checkout of your code etc

I imagine we would use the node AWS SDK to talk to AWS fargate. This might be through keys or IAM role. We would want an option for either, as if we ran this on our Jenkins instance we would want to use IAM for security reasons. 

I think there would be a build step related to building the container. Given we have many projects we would have a default container defined and then the build step would copy the code for a given project and push the generated container to a registry. 

So wdio config would also need config for the container registry and the container. Then fargate would execute the command in the container which has the dependencies and a checkout of the code.

obviously wdio would want as generic an implementation as possible while supporting common use cases. hopefully in the next couple weeks i'll get to try fargate and spec what this would look like

**Comment by christian-bromann** [comment]
> Lambda has limitation on the size of what you can pass to it (which could possibly be made to work but may be harder and less versatile)

I am not sure if we would hit that limitation. I would rather not include any browser driver to it as it increases complexity. I thought that this option would only be used combined with an external grid (e.g. Sauce Labs or home grown grid).

The use case of running the whole test including browser driver within a container should be exported into a service that requires a more complex configuration.

**Comment by dguzmanSM** [comment]
Another issue is that Lambda has a 5 minute limit, our team was also thinking Fargate would be a better solution untill AWS increase Lambda execution time.

**Comment by christian-bromann** [comment]
> Lambda has a 5 minute limit

Oh right, well this pretty much rejects Lambda as an option for running e2e tests

**Comment by emilbader** [comment]
I would like to see https://github.com/webdriverio/webdriverio/issues/2406 in v5.

**Comment by erwinheitzman** [comment]
I think it would be great to add a list of requirements to run all the unit tests.
For example currently I am missing information about you needing to globally install bower, you can't run the mobile tests out of the box either.

It would also be nice if a simple npm install would install all the requirements for running the tests if possible.

**Comment by christian-bromann** [comment]
@erwinheitzman as you might have seen we rebuild the complete [project infrastructure](https://github.com/webdriverio/v5) which will make stuff easier

**Comment by mhemesath** [comment]
I have a feature request idea. Run script injection through babel matching the env the script will be injected to.  This will allow me to do this with a hypothetical example:
```
 var result = browser.execute((...numbers) =>  {
   const result = numbers.reduce((total, value) => total + value);
   return result;
 }, 1, 2, 3, 4);
```

instead of

```
var result = browser.execute(function(a,b,c,d) {
  var result =  a + b + c + d;
  return result;
}, 1, 2, 3, 4);
```

**Comment by christian-bromann** [comment]
@mhemesath sounds like a good idea .. in general we should automatically compile test files and injected script with babel if it is setup .. similar to what Jest is doing

**Comment by mhemesath** [comment]
Idea: Throw a warning if webdriver.io is configured with a  ```maxInstances``` higher than the hubs ```maxSession``` or ```slotCounts.total```.    It seems exceeding either of these can lead to errors or timeouts. 

This information should be possible to get through the hub API:
```
curl http://localhost:4444/grid/api/hub
{  
   "success":true,
   "capabilityMatcher":"org.openqa.grid.internal.utils.DefaultCapabilityMatcher",
   "newSessionWaitTimeout":300000,
   "throwOnCapabilityNotPresent":false,
   "registry":"org.openqa.grid.internal.DefaultGridRegistry",
   "cleanUpCycle":5000,
   "custom":{  

   },
   "host":"10.0.0.3",
   "maxSession":10,
   "servlets":[  

   ],
   "withoutServlets":[  

   ],
   "browserTimeout":100000,
   "debug":false,
   "jettyMaxThreads":-1,
   "port":4444,
   "role":"hub",
   "timeout":90000,
   "enablePassThrough":true,
   "newSessionRequestCount":0,
   "slotCounts":{  
      "free":0,
      "total":5
   }
}
```
Being services seem to be the standard way of integrating a hub configuration into wdio; part of me thinks that the maxInstances configuration should actually come from the service providing the selenium/browser integration.  Standalone Selenium would probably only do 1-5 sessions, whereas a Hub service would execute the API above to configure this dynamically. I'm also sure saucelabs has a way to give back the number of session that can be executed.

**Comment by tavares111** [comment]
I would love to have
#2208

**Comment by christian-bromann** [comment]
@tavares111 please no bugs, only features that don't exist yet

**Comment by patrick-romo** [comment]
It would be nice if we could have a logging option to only see user created logs. Currently if I use `browser.logger.info('Test')`, I can only see the output in conjunction with the RESULT or COMMAND logs. It would be helpful for basic debugging to only see the user-defined logs.

**Comment by christian-bromann** [comment]
@dguzmanSM @kevinmcdonnell getting back to the Lambda vs Fargate idea. I think a 5min limit is fine. By running more stuff in parallel the execution time of the actual spec file decreases and should not exceed 5min.
The fact that it requires to build a container for Fargate is problematic. It should be as easiest as possible to run something on a remote machine. Having Docker as a dependencies is not really "easy".

**Comment by patthiel** [comment]
@christian-bromann @kevinmcdonnell since it looks like we're going the lambda function route, i figured it'd be good to point out that Google has a similar (read: _competing_) product to Lambda called [Google Cloud Functions](https://cloud.google.com/functions/).

It'd be great if we could have service plugins to support both. Perhaps something to investigate after the initial lambda service is complete.

**Comment by christian-bromann** [comment]
@patthiel yes, it will! I will build it on top of [serverless](https://serverless.com) which will support a variety of cloud provider

**Comment by elkemper** [comment]
guys what about make `scroll()` returning web-element? 
it'll be useful in "pipelines" like this
```
$('div.some.classes').scroll().click();
```

**Comment by christian-bromann** [comment]
Just fyi: I just came across to implement this and since I've seen a lot of people doing this I just figured to embed it natively. So in a lot of cases you have this behavior:

```js
const elem = $("selector")
elem.waitForVisible()
elem.click()
```

In v5 we will do the waiting for you. It checks if the initial query did found an element and if not, it would wait for a specific time (configurable in options).

**Comment by klamping** [comment]
@christian-bromann that's a really great idea, as I see this pattern all over the place. 👍

**Comment by MarianoSalerno** [comment]
[Capybara](https://github.com/teamcapybara/capybara) does something like that with the find method, it waits for a configurable time period for an element to appear and, if it isn't found, throws an exception. It also fails if it finds more than one element that matches the selector

**Comment by kevinmcdonnell** [comment]
I think it would be good to have some more logging by default. I'm thinking for each run include:

browser and version (I just lost some time to this as my ci env was running different version to my local dev env)
path to logs (selenium/chromedriver etc) might aid discoverability of these logs for newcomers

there's probably more that is sufficiently succinct that they could be displayed to save time instead of having to rerun tests with `verbose`

**Comment by aamorozov** [comment]
I would love to see the support for Jest as a test runner out of the box. If this will support separation for unit/webdriver tests that'd be awesome. 

Also, better support for docker/docker-compose with selenium grid would be nice since it seems like more and more projects are getting containerized.

**Comment by christian-bromann** [comment]
@aamorozov what does the jest runner provide for e2e testing? It is an excellent runner for unit tests but not really useful for e2e test. Where do you define the capabilities and other options? I recommend to use wdio with jasmine that gives you almost the same assertions as with jest.

**Comment by aamorozov** [comment]
@christian-bromann 
The only reason I've had in mind for that is that our dev team is using jest for unit tests, and it would be very convenient to find a way(if it can be implemented)  to use jest for e2e tests as well. I agree though that there are certain complications, but i saw you've already made some work here https://github.com/christian-bromann/jest-environment-wdio, and i wonder if it is something you would like to include to **v5**.

**Comment by christian-bromann** [comment]
@aamorozov I was just investigating how an integration could possibly look like but as I mentioned above there is no value in adding jest as test runner to run e2e test. It will be more likely that I will implement a jest environment to run jest tests in the browser like karma does. If you want to have jest like assertions, use the wdio with the jasmine framework adapter.

**Comment by christian-bromann** [comment]
Closing this. Thank you very much all for your feedback and input. If you have any other feature request, just create a normal issue on GitHub. Thank you!

**Comment by pradeepvish1213** [comment]
How to open multiremote  with multiple urls.
